micro_dl/preprocessing/resize_images.py

```diff
@@ -107,3 +113,66 @@ def resize_frames(self):
             os.path.join(self.resize_dir, "frames_meta.csv"),
             sep=',',
         )
+
+    def resize_volumes(self, num_slices_subvolume=-1):
+        """Down or up sample volumes
+
+        :param int num_slices_subvolume: num of 2D slices to include in each
+         volume. if -1, include all slices
+        """
+
+        # assuming slice_ids will be continuous
+        num_total_slices = len(self.slice_ids)
+        if not isinstance(self.scale_factor, float):
+            sc_str = '-'.join(self.scale_factor.astype('str'))
+        else:
+            sc_str = self.scale_factor
+
+        mp_args = []
+        resized_metadata_list = []
+        if num_slices_subvolume == -1:
+            num_slices_subvolume = len(self.slice_ids)
+        num_blocks = np.ceil(
+            num_total_slices / num_slices_subvolume
+        ).astype('int')
+        for time_idx in self.time_ids:
+            for pos_idx in self.pos_ids:
+                for channel_idx in self.channel_ids:
+                    for slice_idx in range(num_blocks):
+                        start_idx = slice_idx * num_slices_subvolume
+                        end_idx = start_idx + num_slices_subvolume
+                        if end_idx >= num_total_slices:
+                            end_idx = num_total_slices
+                            start_idx = num_total_slices - num_slices_subvolume
+                        op_fname = 'im_c{}_z{}-{}_t{}_p{}_sc{}.npy'.format(
+                            channel_idx,
+                            start_idx,
+                            end_idx,
+                            time_idx,
+                            pos_idx,
+                            sc_str)
+                        write_fpath = os.path.join(self.resize_dir, op_fname)
+                        mp_args.append((time_idx,
+                                        pos_idx,
+                                        channel_idx,
+                                        start_idx,
+                                        end_idx,
+                                        self.frames_metadata,
+                                        write_fpath,
+                                        self.scale_factor,
+                                        self.input_dir))
+                        cur_metadata = {'time_idx': time_idx,
+                                        'pos_idx': pos_idx,
+                                        'channel_idx': channel_idx,
+                                        'slice_start_idx': start_idx,
+                                        'slice_end_idx': end_idx,
+                                        'file_name': op_fname}
+                        resized_metadata_list.append(cur_metadata)
+
+        # Multiprocessing of kwargs
+        mp_utils.mp_rescale_vol(mp_args, self.num_workers)
+        resized_metadata_df = pd.DataFrame.from_dict(resized_metadata_list)
+        resized_metadata_df.to_csv(
```

Comment:
Won't there be a problem when frames_meta has different fields than expected in the rest of the processing steps like tiling and masks?

Reply:
very true! I'll replace slice_start_idx with start_idx and remove slice_end_idx. We should be able to get the number of tiles per subvolume from the config/json.